Feature 7: Apply to Group Use case (without screens)

ipekakyol opened the pull request:
TODO: create screens and check implementation of current user

soheegoo committed:
First initial version of the Group Creation Use Case. Just wanted to commit to show you guys what I've done so far. Probably more things to change/add

soheegoo committed:
Added main file which can open up Java Swing UI for my Group Register Screen. However, there's no data persistence yet and clicking the buttons don't do anything. Confused on how to switch screens.

aaryavatsa committed:
LeaveGroup use case ... unfinished

ipekakyol committed:
Added the main functions of Feature 7: Apply to Group Use case
TODO: create screens and check implementation of current user

ipekakyol committed:
Added the main functions of Feature 7: Apply to Group Use case
TODO: create screens and check implementation of current user

ipekakyol committed:
Rebasing

ipekakyol committed:
Added the main functions of Feature 7: Apply to Group Use case
TODO: create screens and check implementation of current user

SharonHsieh22 reviewed (commented):
Code looks great! Follows clean architecture and seems functional. I think some edge case errors shouldn't be displayed to the user with prepareFailView but this is a problem Aarya and I are also having so we can discuss how we want to deal with those on Monday.

aaryavatsa committed:
LeaveGroup use case (UI unfinished)

aaryavatsa reviewed (dismissed):
Yeah, I agree with Sharon -- let's dedicate time on Monday (or even after for office hours) to discuss the duplicated DsGateway methods as well as figuring out what to do with the numerous error messages. 

But overall, well done! Nice organizaiton.

ileuxali reviewed (dismissed):
looks good fam

ileuxali committed:
Made 2 changes to User class:
1. made it an interface to allow multiple types of users, currently just normal user
2. deleted userID attribute, that attribute is giving us a lot of difficulty implementing, instead now we assume username attribute is unique

Also finished the implementation of the singleton class CurrentUser following Sohee's started code to store the current logged in user. Also included a test file to showcase how to use the singleton class.

ileuxali committed:
I messed up earlier, the hashmap for groups should be from String to Group not Integer to Group :^)

aaryavatsa committed:
Merge branch 'main' into feature-1-user-signuplogin

ileuxali committed:
Update Group.java

ileuxali committed:
Create ObjectMap.java

ileuxali committed:
Merge pull request #20 from CSC207-2022F-UofT/feature-1-user-signuplogin

Made changes to user and completed singleton class

ipekakyol committed:
Edited the use case methods to be based on username and groupname instead of userID and groupID

ipekakyol committed:
Edited the use case methods to be based on username and groupname instead of userID and groupID

ipekakyol committed:
Edited the use case methods to be based on username and groupname instead of userID and groupID

ipekakyol committed:
Edited the use case methods to be based on username and groupname instead of userID and groupID

aaryavatsa committed:
I've updated my code from the comments that were left, as well as reflected the changes we decided on during tutorial. Both use cases are pretty much finished -- I'm working on the screens and testing right now.

aaryavatsa committed:
Merge branch 'main' into 3-feature-3-leave-group-or-cancel-application

aaryavatsa committed:
Merge branch 'main' of https://github.com/CSC207-2022F-UofT/course-project-groupedup into 3-feature-3-leave-group-or-cancel-application

# Conflicts:
#	src/main/java/Entities/Group.java
#	src/main/java/Entities/User.java

aaryavatsa committed:
Merge remote-tracking branch 'origin/3-feature-3-leave-group-or-cancel-application' into 3-feature-3-leave-group-or-cancel-application

# Conflicts:
#	src/main/java/Entities/Group.java
#	src/main/java/Entities/User.java

soheegoo committed:
Added main file which can open up Java Swing UI for my Group Register Screen. However, there's no data persistence yet and clicking the buttons don't do anything. Confused on how to switch screens.

soheegoo committed:
Added main file which can open up Java Swing UI for my Group Register Screen. However, there's no data persistence yet and clicking the buttons don't do anything. Confused on how to switch screens.

soheegoo committed:
Group Creation Use Case. Including serialization and test files for serialization. Didn't include GroupProfile into the Group Entity because I need Julia's GroupProfile Entity first, or else it give me an error since the class doesn't exist yet.

ileuxali committed:
deleted a useless test

Deleted a useless test. It was made during initial testing stages to show progress. It should've been deleted long ago.

ileuxali committed:
Merge pull request #22 from CSC207-2022F-UofT/1-feature-4-group-creation

1 feature 4 group creation

aaryavatsa committed:
Merge branch 'main' of https://github.com/CSC207-2022F-UofT/course-project-groupedup into 3-feature-3-leave-group-or-cancel-application

# Conflicts:
#	src/main/java/Entities/Group.java

soheegoo commented:
General comment about project organization structure, did you put your use case into a package that should be inside of the src code package? For other people's use cases I see another package within the src/main/java package but yours seems to be in just one. Just making sure so you're not accidentally working in main or something!

soheegoo commented on src/main/java/ApplyToGroup/ApplyToGroupRequestModel.java:
```diff
@@ -0,0 +1,20 @@
+    public String getGroupName() {
+        return groupName;
+    }
+
```
This request model is looking really similar to mine. Maybe we should consider sharing, or having an interface that we can share? Not really sure because we have seperate use cases which should stay seperated for good organization but we can talk about that

aaryavatsa replied:
Yeah maybe it would be good to keep the request models seperate, even if they have similar parameters. I think the majority of the use cases have similar parameters but the response model for each use case should be totally isolated to that specific use case only.

aaryavatsa replied:
Oh, also. We've implemented the current user class, so we're passing around the current user username as well.

ipekakyol replied:
Hello! I think that we should also keep the request models separate. If we can figure out a way that does not impact the organization negatively, we can discuss it before Phase 5.

soheegoo commented on src/main/java/ApplyToGroup/ApplyToGroupResponseModel.java:
```diff
@@ -0,0 +1,31 @@
+
+        this.user = user;
+        this.group = group;
+    //}
```
Just wondering, why are you passing user in the response model?

aaryavatsa replied:
Yes, I don't think we should be passing around entities (even if they're now interfaces), right? Especially in the response model, as the view only needs the username string and the groupname string or any other basic descriptors.

ipekakyol replied:
Yeah I also noticed this after we decided to work with usernames instead of userIDs. I made the necessary changes and they will be reflected in my new pull request.

soheegoo commented on src/main/java/ApplyToGroup/ApplyToGroupInteractor.java:
```diff
@@ -0,0 +1,58 @@
+        if (user.getGroups().containsValue(group)) {
+            return applyToGroupPresenter.prepareFailView("Group is already in user's list.");
+        }
+
```
Great job with all these corner cases! Maybe you could condense some of these if statements by checking multiple conditions in one if statement since you are using the same FailView statement for some of them to make the code cleaner

aaryavatsa replied:
Of course, there is always a balance between concise and readable. Combining checks might make less lines of code but I think what you've done for now is very easy to follow. We can cut down on the length by removing some redundant checks -- reference mine and let me know if you think I've condensed TOO much. We have a very similar structure.

ipekakyol replied:
Sure, I will look at yours and can restructure mine if it works better or review yours as well

soheegoo commented on src/main/java/ApplyToGroup/ApplyToGroupController.java:
```diff
@@ -0,0 +1,18 @@
+
+        return applyToGroupInput.create(requestModel);
+    }
+
```
I believe this should be in the package with your screens to abide by clean architecture because the controller is not in the use case layer

aaryavatsa replied:
Sohee is correct, but it seems you don't have a screens package just yet. That's totally fine, it's not hard at all to transfer classes around different packages as intellij refactors everything for you very smoothly.

ipekakyol replied:
Yes, thank you aarya, i had not worked on the screens package just yet.

aaryavatsa committed:
- Removed some test cases in my interactors (checking whether current user exists is useless!) and by extension, the DsGateways
- Added userInGroup() and groupInUser() methods in DsGateways (for edge cases and testing)
- * RENAMED Presenter interface -> OutputBoundary interface and ResponseFormatter class -> Presenter class (Paul's example has the files named incorrectly, he clarified this in lec and it would be good for us to rename our files to the correct name) *
- Added some methods in the entity classes regarding my use cases
- Fixed some warnings for Leo and Sohee's interfaces (we don't need the keyword public)

- Presenter class is currently under construction
- I lost my testing code :( git stash fail. I'll rewrite that and have that up soon

aaryavatsa committed:
Javadocs and method docs!

aaryavatsa commented on src/main/java/ApplyToGroup/ApplyToGroupInteractor.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @Override
+    public ApplyToGroupResponseModel create(ApplyToGroupRequestModel requestModel) {
```
Let's actually rename this method to be more descriptive (as we talked about in our meeting, not sure if you've already followed up on this). create() was just the method Paul used in his usecase. For yours it should be referential to the use case you are coding -- maybe something like applyToGroup() instead.

ipekakyol replied:
Yes, I changed it after our conversation, I will update shortly.

aaryavatsa commented on src/main/java/ApplyToGroup/ApplyToGroupInteractor.java:
```diff
@@ -0,0 +1,58 @@
+
+    @Override
+    public ApplyToGroupResponseModel create(ApplyToGroupRequestModel requestModel) {
+        if (!applyToGroupDsGateway.userExistsByName(user.getUsername())) {
```
Ah, this is the same issue I had. Please read the discussion that Hannah and I had in the comments for my pull request. I'm not sure if this check is needed!

aaryavatsa commented on src/main/java/ApplyToGroup/ApplyToGroupInteractor.java:
```diff
@@ -0,0 +1,58 @@
+        }
+
+        if (group.getGroupLeader().getUsername().equals(user.getUsername())) {
+            return applyToGroupPresenter.prepareFailView("User is group leader.");
```
Some of these checks seem trivial.. please correct me if I'm missing something. It seems you performed the same checks that I did in my usecase, but our usecases are not THAT similar. How would we ever get a situation where a user can click the "apply to group" button if the user is the group leader? The view or screen should never be in that position and displaying that button. Let's get together and discuss this before our next meeting. This line of thought is what made me delete some of my own checks too.

ipekakyol replied:
I wanted to make sure that I didn't miss anything but it also does not make sense to me. I think I just did not have enough information or understanding of the screens yet so put any checks I could with the intention of deleting it later.

aaryavatsa commented on src/main/java/ApplyToGroup/ApplyToGroupPresenter.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface ApplyToGroupPresenter {
+
+    ApplyToGroupResponseModel prepareSuccessView(ApplyToGroupResponseModel user);
```
I think you should rename the parameter to be more descriptive, maybe add a doc string. The response model is carrying info about the user and group, right?

aaryavatsa reviewed (changes requested):
Overall, I think this was a great start -- it's been a while since your last commit to this branch but I know you've been working on it since then. I think it would be good for you to commit your latest changes, as I'm sure some of the things we've reviewed have already been fixed by you. We can review the new changes.

ipekakyol committed:
Added output boundary, moved controller and presenter to screens package (still being updated), made changes to be more concise

ipekakyol committed:
Added output boundary, moved controller and presenter to screens package (still being updated), made changes to be more concise

ipekakyol committed:
Merge remote-tracking branch 'origin/11-feature-7-applying-to-group' into 11-feature-7-applying-to-group

# Conflicts:
#	src/main/java/ApplyToGroup/ApplyToGroupDsGateway.java
#	src/main/java/ApplyToGroup/ApplyToGroupInputBoundary.java
#	src/main/java/ApplyToGroup/ApplyToGroupInteractor.java
#	src/main/java/ApplyToGroup/ApplyToGroupRequestModel.java
#	src/main/java/ApplyToGroup/ApplyToGroupResponseModel.java

ipekakyol committed:
Added output boundary, moved controller and presenter to screens package (still being updated), made changes to be more concise

ipekakyol committed:
Added output boundary, moved controller and presenter to screens package (still being updated), made changes to be more concise

aaryavatsa committed:
Merge pull request #17 from CSC207-2022F-UofT/3-feature-3-leave-group-or-cancel-application

LeaveGroup and CancelApplication use cases

ipekakyol committed:
rebasing

ipekakyol committed:
Committing to avoid merge conflicts, will update later

ipekakyol committed:
* Initialized usermap to be able to pass it as an argument for getMemberRequests() which was edited to have the usermap parameter in the Group entity
* Added ApplyToGroupScreen and ApplytoGroupFailed files
* Added javadocs following Aarya and Sharon's formatting

ipekakyol committed:
Merge remote-tracking branch 'origin/11-feature-7-applying-to-group' into 11-feature-7-applying-to-group

# Conflicts:
#	src/main/java/ApplyToGroup/ApplyToGroupDsGateway.java
#	src/main/java/ApplyToGroup/ApplyToGroupInteractor.java
#	src/main/java/ApplyToGroup_screens/ApplyToGroupController.java

ipekakyol committed:
* Deleted unnecessary copies

ipekakyol commented:
Making a new pull request on November 19th to have a clearer picture of the files that have been changed. Closing this pull request to avoid conflicts
